pkg/apis/keda/v1alpha1/triggerauthentication_types.go

```diff
@@ -83,6 +92,157 @@ type AuthEnvironment struct {
+	// +optional
+	Token string `json:"token"`
+
+	// +optional
+	ServiceAccount string `json:"serviceAccount"`
+}
+
+// VaultAuthentication contains the list of Hashicorp Vault authentication methods
+type VaultAuthentication string
+
+// Client authenticating to Vault
+const (
+	VaultAuthenticationToken      VaultAuthentication = "token"
+	VaultAuthenticationKubernetes                     = "kubernetes"
+	// VaultAuthenticationAWS                            = "aws"
+)
+
+// VaultSecret defines the mapping between the path of the secret in Vault to the parameter
+// +k8s:openapi-gen=true
+type VaultSecret struct {
+	Parameter string `json:"parameter"`
+	Path      string `json:"path"`
+	Key       string `json:"key"`
+}
+
 func init() {
 	SchemeBuilder.Register(&TriggerAuthentication{}, &TriggerAuthenticationList{})
 }
+
+// Authenticate returns a client connected to Vault
+func (v *Vault) Authenticate(logger logr.Logger) (*vaultApi.Client, error) {
+	config := vaultApi.DefaultConfig()
+	client, err := vaultApi.NewClient(config)
+
+	err = client.SetAddress(v.Address)
+	if err != nil {
+		return nil, err
+	}
+
+	token, err := v.token(client)
+	if err != nil {
+		return nil, err
+	}
+	client.SetToken(token)
+
+	lookup, err := client.Auth().Token().LookupSelf()
+	//If token is not valid so get out of here early
+	if err != nil {
+		return nil, err
+	}
+
+	renew := lookup.Data["renewable"].(bool)
+	if renew == true {
+		go v.renewToken(client, logger)
+	}
+
+	return client, nil
+}
+
+func (v *Vault) token(client *vaultApi.Client) (string, error) {
+	var token string
+
+	switch v.Authentication {
+	case VaultAuthenticationToken:
+		// Got token from VAULT_TOKEN env variable
+		if len(client.Token()) > 0 {
+			break
+		} else if len(v.Credential.Token) > 0 {
+			token = v.Credential.Token
+		} else {
+			return token, errors.New("Could not get Vault token")
+		}
+	case VaultAuthenticationKubernetes:
+		if len(v.Mount) == 0 {
+			return token, errors.New("Auth mount not in config")
+		}
+
+		if len(v.Role) == 0 {
+			return token, errors.New("K8s role not in config")
+		}
+
+		if len(v.Credential.ServiceAccount) == 0 {
+			return token, errors.New("K8s SA file not in config")
+		}
+
+		//Get the JWT from POD
+		jwt, err := ioutil.ReadFile(v.Credential.ServiceAccount)
+		if err != nil {
+			return token, err
+		}
+
+		data := map[string]interface{}{"jwt": string(jwt), "role": v.Role}
+		secret, err := client.Logical().Write(fmt.Sprintf("auth/%s/login", v.Mount), data)
+		if err != nil {
+			return token, err
+		}
+
+		token = secret.Auth.ClientToken
+	default:
+		return token, fmt.Errorf("Vault auth method %s is not supported", v.Authentication)
+	}
+
+	return token, nil
+}
+
+func (v *Vault) renewToken(client *vaultApi.Client, logger logr.Logger) {
+	secret, err := client.Auth().Token().RenewSelf(0)
+	if err != nil {
+		logger.Error(err, "Vault renew token: failed to create the payload")
+	}
+
+	renewer, err := client.NewRenewer(&vaultApi.RenewerInput{
```

Comment:
how is the lifetime for this renewer managed? Is there a way to signal cancellation to it for cleanup

Reply:
fixed with channel and stop method in hashicorpvault_handler.go